### PR TITLE
Issue 5690: Adding new stream or rg or kvt to scope should lazily create the metadata table if it doesnt exist

### LINKED ISSUE #5690: Post create scope, lazily create streamsinscopeTable, rgsInScopeTable and kvtsInScopeTable and delete them with delete scope
**Problem description**
CreateScope is not implemented as a failover protected workflow. Which means we could have failed after adding the entry to scopes table but without creating corresponding streamsInScope table, rgsInScope table and kvtsInScope table.

These should be created lazily if they do not exist if an attempt to add streams/rgs/kvts to scope is attempted. 

Similarly, at the time of delete scope, these tables should be removed. Presently we only remove streamsInScopeTable. 

**Problem location**
PravegaTablesScopes

**Suggestions for an improvement**
Fix the bugs.

### PR DESCRIPTION
**Pull request by shiveshr**
Signed-off-by: Shivesh Ranjan <shivesh.ranjan@gmail.com>

**Change log description**  
- Post create scope, lazily create streamsinscopeTable, rgsInScopeTable and kvtsInScopeTable if they dont exist
- delete scope should not be allowed if rg table and kvt table are non empty

**Purpose of the change**  
Fixes #5690 

**What the code does**  
In PravegaTablesScopes we create tables for stream, kvt, and rg but that is not done as a workflow. So after adding the entry we could have failed. This could make the scope unusable unless createScope is called again.
The way we fix it is, any access to the scope (for add`resource` or list`resource`) should not result in datanotfound if the entry in scopes table was successfully created. 
So we lazily create table as part of addStream, addReaderGroup and addKvt if they dont exist.
and we return empty list if tables dont exist. 

Secondly, deletescope should be predicated on whether all resources under the scope have been cleaned up or not. 

**How to verify it**  
Unit test added

### CODE REVIEW

File: controller/src/main/java/io/pravega/controller/store/PravegaTablesScope.java [hunk 1]
```diff
@@ -69,14 +71,8 @@ public String getName() {
...
         // If scope entry exists in Scopes table, create the streamsInScope table before throwing DataExists exception
-        return Futures.handleCompose(Futures.exceptionallyComposeExpecting(storeHelper.addNewEntry(
-                SCOPES_TABLE, scopeName, newId()),
-                DATA_NOT_FOUND_PREDICATE,
-                () -> storeHelper.createTable(SCOPES_TABLE)
-                                 .thenCompose(v -> {
-                                     log.debug("table for scopes created {}", SCOPES_TABLE);
-                                     return storeHelper.addNewEntryIfAbsent(SCOPES_TABLE, scopeName, newId());
-                                 })), (r, e) -> {
+        return Futures.handleCompose(withCreateTableIfAbsent(() -> storeHelper.addNewEntry(
```
Comment (pbelgundi) [inline controller/src/main/java/io/pravega/controller/store/PravegaTablesScope.java:74]: withCreateTableIfAbsent() can also be used on Streams, ReaderGroups and KVTs

File: controller/src/main/java/io/pravega/controller/store/PravegaTablesScope.java [hunk 2]
```diff
@@ -140,37 +136,35 @@ public String getName() {
...
+                            storeHelper.deleteTable(kvtsInScopeTableName, true), 
+                            storeHelper.deleteTable(rgsInScopeTableName, true))
+                                     .thenAccept(v -> log.debug("tables deleted {} {} {}", streamsInScopeTableName,
+                                             kvtsInScopeTableName, rgsInScopeTableName));
+                })
                 .thenCompose(deleted -> storeHelper.removeEntry(SCOPES_TABLE, scopeName));
     }
 
     @Override
     public CompletableFuture<Pair<List<String>, String>> listStreams(int limit, String continuationToken, Executor executor) {
```
Comment (pbelgundi) [inline controller/src/main/java/io/pravega/controller/store/PravegaTablesScope.java:157]: We also need a listReaderGroups() API for listing all ReaderGroups inside a scope.

**Review by pbelgundi — APPROVED** [review]
LGTM with some minor comments.

File: controller/src/main/java/io/pravega/controller/store/PravegaTablesScope.java (hunk 2 above)
Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/store/PravegaTablesScope.java:157]: we presently have a rest api for listing readergroups in a scope.. it does so by listing all streams in a scope and then filtering those with prefix _RG. 
Perhaps that should be changed to a new api in streammetadatastore which lists them from the rg metadata.. just that it wont list any older rgs.

File: controller/src/main/java/io/pravega/controller/store/PravegaTablesScope.java (hunk 1 above)
Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/store/PravegaTablesScope.java:74]: we use it dont we?

Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/store/PravegaTablesScope.java:74]: yes we do.

**Pull request merged**